tower-http/src/services/fs/serve_dir.rs

```diff
@@ -251,38 +301,45 @@ impl Future for ResponseFuture {
-                    Ok(Output::File(file, mime, chunk_size)) => (file, mime, chunk_size),
-
-                    Ok(Output::Redirect(location)) => {
-                        let res = Response::builder()
-                            .header(http::header::LOCATION, location)
-                            .status(StatusCode::TEMPORARY_REDIRECT)
-                            .body(empty_body())
-                            .unwrap();
-                        return Poll::Ready(Ok(res));
-                    }
-
-                    Ok(Output::NotFound) => {
-                        let res = Response::builder()
-                            .status(StatusCode::NOT_FOUND)
-                            .body(empty_body())
-                            .unwrap();
-
-                        return Poll::Ready(Ok(res));
-                    }
-
-                    Err(err) => {
-                        return Poll::Ready(
-                            super::response_from_io_error(err).map(|res| res.map(ResponseBody)),
-                        )
-                    }
-                };
+                let (file, mime, chunk_size, max_age) =
+                    match ready!(Pin::new(open_file_future).poll(cx)) {
+                        Ok(Output::File(file, mime, chunk_size, max_age)) => {
+                            (file, mime, chunk_size, max_age)
+                        }
+
+                        Ok(Output::Redirect(location)) => {
+                            let res = Response::builder()
+                                .header(http::header::LOCATION, location)
+                                .status(StatusCode::TEMPORARY_REDIRECT)
+                                .body(empty_body())
+                                .unwrap();
+                            return Poll::Ready(Ok(res));
+                        }
+
+                        Ok(Output::NotFound) => {
+                            let res = Response::builder()
+                                .status(StatusCode::NOT_FOUND)
+                                .body(empty_body())
+                                .unwrap();
+
+                            return Poll::Ready(Ok(res));
+                        }
+
+                        Err(err) => {
+                            return Poll::Ready(
+                                super::response_from_io_error(err).map(|res| res.map(ResponseBody)),
+                            )
+                        }
+                    };
                 let body = AsyncReadBody::with_capacity(file, chunk_size).boxed();
                 let body = ResponseBody(body);
 
                 let mut res = Response::new(body);
                 res.headers_mut().insert(header::CONTENT_TYPE, mime);
+                res.headers_mut().insert(
+                    header::CACHE_CONTROL,
+                    HeaderValue::from_str(&format!("max-age={}", max_age.as_secs())).unwrap(),
```

Comment:
To avoid allocating a string for each response (only to create a `HeaderValue`) I think should instead change the field from a `Duration` (or rather `Option<Duration>`) to `Option<HeaderValue>` and then generate the header value directly when the user sets it. Cloning a `HeaderValue` is cheaper since its reference counted.